README.rst
===================================
YADE - Yet Another Dynamic Engine
===================================

Extensible open-source framework for discrete numerical models, focused on 
Discrete Element Method. The computation parts are written in c++ using flexible
object model, allowing independent implementation of new algorithms and interfaces. 
Python is used for rapid and concise scene construction, simulation control, 
postprocessing and debugging.

**Documentation:**

- https://www.yade-dem.org/doc/ (includes c++ and python reference)
- https://yade-dem.org/doc/Yade.pdf (PDF-file with documentation)
- https://yade-dem.org/doc/Yade.epub (EPUB-file with documentation)
 
**Downloads:**

- Packages for Ubuntu and Debian:
  
  - https://yade-dem.org/doc/installation.html#packages
  - External packages: https://launchpad.net/~yade-users/+archive/external

- Tarballs of different versions: 

  - https://gitlab.com/yade-dev/trunk/tags
  - Repository on gitlab.com: https://gitlab.com/yade-dev/trunk/-/archive/master/trunk-master.tar.gz
  - Repository on git.debian.org: http://git.debian.org/?p=debian-science/packages/yade.git

**Teams involved:**

- General users: https://launchpad.net/~yade-users
- Developers and advanced users: https://launchpad.net/~yade-dev
- Developers for distributed memory parallelization of Yade: https://launchpad.net/~yade-mpi
- Infrastructure maintainers: https://launchpad.net/~yade-pkg

**Miscellaneous:**

- `Wiki <https://yade-dem.org/wiki/>`_
- `Screenshots <https://www.yade-dem.org/wiki/Screenshots_and_videos>`_
- `Tutorial <https://yade-dem.org/doc/tutorial-examples.html>`_
- `Videos <https://yade-dem.org/doc/tutorial-more-examples-fast.html>`_
- `Yade page on Open Hub <http://www.openhub.net/p/yade/>`_
- `Yade on Debian PTS <http://packages.qa.debian.org/y/yade.html>`_
- `Yade on Debian salsa <https://salsa.debian.org/science-team/yade/commits/master>`_
- `Yade in Ubuntu <https://launchpad.net/ubuntu/+source/yade>`_
- `Launchpad page <http://www.launchpad.net/yade>`_


=========================================================================================
YADE - Yet Another Dynamic Engine - Python 3 branch - remove this when merge is complete.
=========================================================================================

The packages needed are the python3-versions of all packages listed in the official Yade doc. So the installation command becomes:

.. code::

sudo apt install cmake git freeglut3-dev libloki-dev libboost-all-dev fakeroot dpkg-dev build-essential g++ python3-dev python3-ipython python3-matplotlib libsqlite3-dev python3-numpy python3-tk gnuplot libgts-dev python3-pygraphviz libvtk6-dev libeigen3-dev python3-xlib python3-pyqt5 pyqt5-dev-tools python3-pyqt5.qtwebkit gtk2-engines-pixbuf python3-pyqt5.qtsvg libqglviewer-dev-qt5 python3-pil libjs-jquery python3-sphinx python3-git libxmu-dev libxi-dev libcgal-dev help2man libbz2-dev zlib1g-dev python3-minieigen libopenblas-dev libsuitesparse-dev libmetis-dev python3-bibtexparser python3-future


On ubuntu 16.04 python3-ipython is to be replaced by ipython3, and libqglviewer-dev-qt5 by libqglviewer-dev.

Additional dependency for backward python 2.7 compatibility: python-future (in addition to all python-xxx packages).

Installing, running, testing, doc look now ok, on all current and future debian/ubuntu versions. You can try it with:

.. code::

cmake -DCMAKE_INSTALL_PREFIX=/path/to/your/install/target -DPYTHON_VERSION=3 /path/to/sources

**Implementation status**

- cmake and yade compilation ok with python3
- yade launch ok with python3, generating/launching triax ok
- yade --test all ok with python3 (-DPYTHON_VERSION=3)
- yade --test all ok with python2 (-DPYTHON_VERSION=2)
- yade --checks almost ok with python2 and python3 (but the only issue is not related to python version)

**Known bugs/todos**

- (DONE) update CMakeLists.txt to make python3 and python2 optional (and one of them at least is necessary)
- (FIXED by Vaclav) pygts may not compile
- (FIXED) python3-bibtex package doesn't exist, and underlying _bibtex python module will not be found. As a result, the doc will not compile. We may use another python bibtex parser or use sphinxcontrib-bibtex: https://sphinxcontrib-bibtex.readthedocs.io/en/latest/usage.html
- (PARTIALLY FIXED) fix checks/tests (checks neither work in the master branch)
- (DONE) try/fix python2 retro-compatibility
- try/fix examples




ChangeLog
==================================================
yade-2021.01a
Sun, Jan 17 18:45:12 2021 +0100

Short summary of 2021.01a release:
short title (Merge Request count)

Documentation                         (41 MRs)
Bug fixes                             (45 MRs)
High precision RealHP<N>              (31 MRs)
Small improvements                    (27 MRs)
Fluid & thermal & particle saturation (15 MRs)
Tests & checks                        (13 MRs)
OpenMPI calculations                  (13 MRs)
clang-format in the pipeline          ( 7 MRs)
Packaging improvements                ( 7 MRs)
GUI                                   ( 3 MRs)


Documentation (41 MRs):
  Update installation.rst (asan builds + stuff)
  Add recent publications
  point youtube's 'related videos' to the same channel
  Update front page link to videos: tutorial-more-examples.html more up to date.
  Fix bug and documentation FlowEngine invVoidVolume wrapper
  Fix recent build failures on master with latest docker bullseye image.
  Small docs fix
  Doc typo: missing norm in trial shear force
  Doc Body.groupMask
  User doc
  Update high precision documentation
  Miscellanea doc
  Add example script for opengeomechanics paper
  clarify thermalengine doc
  Documentation makeCloud and makeClumpCloud
  ViscoelasticPM doc update
  Prog doc typo
  ProgDocTypos
  Formulation doc typo
  improve cellDist doc
  small addition to the doc to mention verletDist explicitely
  Doc prog.rst update
  FlowEngine.getCell doc
  Doc update of VTKRecorder.fileName
  ymport.textPolyhedra-doc
  Prog doc typos and minor changes
  Doc Law2_Polyhedra*
  residualFrictionAngle in radians in JCFpmMat docstring
  Sphinx doc broken link in Prog doc
  O.forces.f doc
  Two Docstrings
  BaselineTypo
  adding open geomechanics reference to doc and pubs
  Add --stdperformance, --quickperformance explanation to documentation.
  Polyhedra Ip2 doc
  MaterialContainer doc typo
  timing.runtimeDoc
  Added Hicham Benniou PhD reference
  Add references to article describing lubrication law
  update publications
  !463 alternative : Added article February 2020 Computational Particle Mechanics

Bug fixes (45 MRs):
  fix #145 seed in makeCloud FIX makeCloud doesn't use "seed"
  Fix pygts module import in python3.9
  Fix PyRunner. Now it can propagate errors upwards.
  Fix #110 FIX Graph yade.plot.live matplotlib numpy update exception
  Fix compilation with python3.9
  fix #157 FIX O.stopAtTime does not do anything useful
  Fix #140 FIX ForceContainerParallel::sync() can count permanent forces twice or more
  Fix https://gitlab.com/yade-dev/docker-yade/issues/1 FIX yade-ci crashes
  export.text2vtk - fixed a bug + refactoring
  Fix #146 FIX "to-Python converter already registered" RuntimeWarning
  Fix even more warnings, print only 'maybe-uninitialized' and '#warning' (!270), faster compilation. (!273)   18 of 18 tasks completed
  fix clang formating mistake, change remaining doubles to Real
  fix tutorial script https://answers.launchpad.net/yade/+question/689537
  Fix printing Vectors by LOG macros.
  Fix GCC-10 compilation #958006
  Fix_PBs_orientations
  #153 fix FIX O.energy['nonViscDamp'] misses gravity forces
  Skip url ping-pong for "getting help" page
  fix yade crash, when closing 3D view via "X" window frame button, in Ubuntu 20.04, fixes #164 FIX yade crash when closing the 3D view.
  Fix close view crash
  small mailmap fix and test cleanup
  Little bit of work on #97 : lib/base/Math.hpp should be math only 
  Fix compilation against newer QT5.
  Some small fixes for Sacha : Fix compilation with boost 1.65 and g++ 9.3 combination
  Fix denser grid in GLViewer.
  clumpId type should be Body::id_t, not int. Add fixme note in Math.hpp about issue #97
  Fix #162 - fix building master pipeline due to recent VTK upgrade in OpenSuse 15.
  Fix build on latest bullseye
  Fix doc generation for newer sphinx. #955097
  Workaround MPFR boost bug on ubuntu focal
  Enable OpenMP for MPFR
  SpatialQuickSortCollider is nice, because it's only 70 lines long. But has one bug. FIX groupMask treatment.
  comment out buggy heat flux boundary condition for now
  GCAL version 5.1.1 produces different results (again), but maybe the old results are back?
  Qt::MidButton is deprecated, use MiddleButton instead
  Use boost::filter_iterator for BodyContainer instead of home-backed. #122
  workaround #183
  Add compilation on multiarch: ppc64el, arm64. Workaround #179
  workaround #178 , use /tmp/SiloLargeOrifice.stl
  Fix #180
  Fix https://gitlab.com/yade-dev/docker-yade/issues/1
  make checkPolyhedraCrush.py single core only
  Disable three tries if there was a "script failure" reason to fail the job
  Fix #174 add --checkall option.
  Fix #150 : OpenMP crash when deleting bodies in yade -j4 ./examples/clumps/clump-hopper-test.py

High precision RealHP<N> (31 MRs):
  Finalize high precision Real - finished, really :) documentation + put minieigen sources into Yade
  Add check test for Runge Kutta Cash Karp 54 time integrator, error tolerance 1e-147, 147 correct decimal places :)
  checkGravity.py : reduce error tolerance from 1e-3 → 1e-10
  Use the minieigen ↔ yade.minieigenHP redirection when necessary.
  Add High Precision to the gitlab CI pipeline
  Ensure that VTK is compatibile with Real - Without Macros.
  std::weibull_distribution can't understand Real type so use boost::weibull_distribution
  OpenGL Real compatibility
  Lapack Compatibility with Real type.
  Remove Real from global namespace.
  Another step towards enabling high precision Real
  Adapt checks and tests to high precision Real.
  Add RealHP<N> : double, quadruple and more precisions.
  Bit correctness: extra tests for RealHP<N>
  Small HP cleanup: don't use std::to_string.
  The last remaining double to Real changes (2/2)
  The last remaining double to Real changes (1/2)
  Remove #DEFINEs from MathFunctions.hpp
  Remove unused variables uN, uT.
  don't use operator++ for floating point type.

py/system.py
# coding: utf-8
# 2009 © Václav Šmilauer <eudoxos@arcig.cz>

"""
Functions for accessing yade's internals; only used internally.
"""
from __future__ import print_function
from builtins import object
import sys
from yade import wrapper
from yade._customConverters import *
from yade import runtime
from yade import config
O=wrapper.Omega()

def childClasses(base,recurse=True,includeBase=False):
	"""Enumerate classes deriving from given base (as string), recursively by default. Returns set."""
	ret=set(O.childClassesNonrecursive(base)); ret2=set();
	if includeBase: ret|=set([base])
	if not recurse: return ret
	for bb in ret:
		ret2|=childClasses(bb)
	return ret | ret2

_allSerializables=childClasses('Serializable')
## set of classes for which the proxies were created
_proxiedClasses=set()

## deprecated classes
# if old class name is used, the new object is constructed and a warning is issued about old name being used
# keep chronologically ordered, oldest first; script/rename-class.py appends at the end
_deprecated={
	'GranularMat':'FrictMat', # Sun Jan 10 09:26:45 2010, vaclav@flux
	'SimpleElasticRelationships':'Ip2_FrictMat_FrictMat_NormShearPhys', # Sun Jan 10 09:28:17 2010, vaclav@flux
	'NormalInteraction':'NormPhys', # Sun Jan 10 09:28:56 2010, vaclav@flux
	'NormalShearInteraction':'NormShearPhys', # Sun Jan 10 09:29:22 2010, vaclav@flux
	'ElasticMat':'ElastMat', # Sun Jan 10 09:53:15 2010, vaclav@flux
	'ElasticContactInteraction':'FrictPhys', # Sun Jan 10 09:57:59 2010, vaclav@flux
	'ef2_Spheres_Elastic_ElasticLaw':'Law2_ScGeom_FrictPhys_Basic', # Sun Jan 10 09:59:42 2010, vaclav@flux
	'Ip2_FrictMat_FrictMat_NormShearPhys':'Ip2_FrictMat_FrictMat_FrictPhys', # Sun Jan 10 10:07:40 2010, vaclav@flux
	'GLDrawCpmPhys':'Gl1_CpmPhys', # Sat Feb  6 14:46:08 2010, vaclav@flux
	'RockJointLawRelationships':'Ip2_2xCohFrictMat_NormalInelasticityPhys', # Mon Feb  8 11:17:59 2010, jduriez@c1solimara-l
	'TetraBang':'TTetraGeom', # Tue Feb  9 10:21:24 2010, vaclav@flux
	'TetraMold':'Tetra', # Tue Feb  9 10:22:15 2010, vaclav@flux
	'TetraAABB':'Bo1_Tetra_Aabb', # Tue Feb  9 10:22:33 2010, vaclav@flux
	'Tetra2TetraBang':'Ig2_Tetra_Tetra_TTetraGeom', # Tue Feb  9 10:23:19 2010, vaclav@flux
	'TetraLaw':'TetraVolumetricLaw', # Tue Feb  9 10:24:10 2010, vaclav@flux
	'DirecResearchEngine':'Disp2DPropLoadEngine', # Wed Mar 10 12:23:42 2010, jduriez@c1solimara-l
	'Ip2_BMP_BMP_CSPhys':'Ip2_2xFrictMat_CSPhys', # Wed Mar 10 15:08:56 2010, eudoxos@frigo
	'NormalInelasticityLaw':'Law2_ScGeom_NormalInelasticityPhys_NormalInelasticity', # Wed Mar 17 15:50:59 2010, jduriez@c1solimara-l
	'CapillaryCohesiveLaw':'CapillaryLaw', # Tue Mar 30 14:11:36 2010, sch50p@fluent-ph
	'SimpleElasticRelationshipsWater':'Ip2_Frictmat_FrictMat_CapillaryLawPhys', # Tue Mar 30 14:20:36 2010, sch50p@fluent-ph
	'CapillaryLaw':'Law2_ScGeom_CapillaryPhys_Capillarity', # Wed Mar 31 09:23:36 2010, sch50p@fluent-ph
	'CapillaryParameters':'CapillaryPhys', # Wed Mar 31 09:25:03 2010, sch50p@fluent-ph
	'Ip2_FrictMat_FrictMat_CapillaryLawPhys':'Ip2_FrictMat_FrictMat_CapillaryPhys', # Wed Mar 31 09:26:04 2010, sch50p@fluent-ph
	'SimpleViscoelasticMat':'ViscElMat', # Fri Apr  9 19:25:38 2010, vaclav@flux
	'SimpleViscoelasticPhys':'ViscElPhys', # Fri Apr  9 19:26:34 2010, vaclav@flux
	'Law2_Spheres_Viscoelastic_SimpleViscoelastic':'Law2_ScGeom_ViscElPhys_Basic', # Fri Apr  9 19:28:02 2010, vaclav@flux
	'Ip2_SimleViscoelasticMat_SimpleViscoelasticMat_SimpleViscoelasticPhys':'Ip2_ViscElMat_ViscElMat_ViscElPhys', # Fri Apr  9 19:28:48 2010, vaclav@flux
	'MomentEngine':'TorqueEngine', # Sun May  2 16:09:34 2010, vaclav@flux
	'JumpChangeSe3':'StepDisplacer', # Sun May  2 16:14:21 2010, vaclav@flux
	'Ip2_2xCohFrictMat_NormalInelasticityPhys':'Ip2_2xNormalInelasticMat_NormalInelasticityPhys', # Fri Jun  4 15:36:41 2010, jduriez@c1solimara-l
	'Ip2_2xCohFrictMat_NormalInelasticityPhys':'Ip2_2xNormalInelasticMat_NormalInelasticityPhys', # Fri Jun  4 15:37:01 2010, jduriez@c1solimara-l
	'Ip2_2xCohFrictMat_NormalInelasticityPhys':'Ip2_2xNormalInelasticMat_NormalInelasticityPhys', # Fri Jun  4 15:37:16 2010, jduriez@c1solimara-l
	'OpenGLRenderingEngine':'OpenGLRenderer', # Sat Jul 24 06:04:13 2010, vaclav@flux
	'PeriodicPythonRunner':'PyRunner', # Wed Sep  1 16:41:50 2010, chia@engs-018373
	'InteractionDispatchers':'InteractionLoop', # Mon Sep 27 13:44:54 2010, chia@engs-018373
	'InteractionGeometry':'IGeom', # Thu Sep 30 10:39:24 2010, chia@engs-018373
	'InteractionPhysics':'IPhys', # Thu Sep 30 10:39:43 2010, chia@engs-018373
	'InteractionGeometryFunctor':'IGeomFunctor', # Thu Sep 30 14:27:43 2010, chia@engs-018373
	'InteractionPhysicsFunctor':'IPhysFunctor', # Thu Sep 30 14:27:53 2010, chia@engs-018373
	'InteractionPhysicsDispatcher':'IPhysDispatcher', # Thu Sep 30 14:28:12 2010, chia@engs-018373
	'InteractionGeometryDispatcher':'IGeomDispatcher', # Thu Sep 30 14:28:22 2010, chia@engs-018373
	'Law2_ScGeom_FrictPhys_Basic':'Law2_ScGeom_FrictPhys_CundallStrack', # Wed Oct 13 17:40:42 2010, bchareyre@dt-rv020
	'Law2_ScGeom_CohFrictPhys_ElasticPlastic':'Law2_ScGeom_CohFrictPhys_CohesionMoment', # Wed Oct 13 17:47:09 2010, bchareyre@dt-rv020
	'Law2_ScGeom_NormalInelasticityPhys_NormalInelasticity':'Law2_ScGeom6D_NormalInelasticityPhys_NormalInelasticity', # Thu Oct 28 18:09:16 2010, jduriez@c1solimara-l
	'SpiralEngine':'HelixEngine', # Sat Oct 30 05:52:06 2010, vaclav@flux
	'InterpolatingSpiralEngine':'InterpolatingHelixEngine', # Sat Oct 30 05:52:21 2010, vaclav@flux
	'SpiralInteractionLocator2d':'HelixInteractionLocator2d', # Sat Oct 30 05:53:04 2010, vaclav@flux
	'Law2_ScGeom_CohFrictPhys_CohesionMoment':'Law2_ScGeom6D_CohFrictPhys_CohesionMoment', # Fri Nov 12 18:45:23 2010, bchareyre@dt-rv020
	'Ig2_ChainedCylinder_ChainedCylinder_ScGeom':'Ig2_ChainedCylinder_ChainedCylinder_ScGeom6D', # Fri Nov 12 18:47:44 2010, bchareyre@dt-rv020
	'Ip2_2xCohFrictMat_CohFrictPhys':'Ip2_CohFrictMat_CohFrictMat_CohFrictPhys', # Tue Dec 21 16:11:28 2010, bchareyre@dt-rv020
	'Ig2_Sphere_Sphere_L3Geom_Inc':'Ig2_Sphere_Sphere_L3Geom', # Sun Dec 26 11:11:05 2010, vaclav@flux
	'Ig2_Wall_Sphere_L3Geom_Inc':'Ig2_Wall_Sphere_L3Geom', # Sun Dec 26 11:11:30 2010, vaclav@flux
	'Ig2_Sphere_Sphere_L6Geom_Inc':'Ig2_Sphere_Sphere_L6Geom', # Sun Dec 26 11:11:53 2010, vaclav@flux
	### END_RENAMED_CLASSES_LIST ### (do not delete this line; scripts/rename-class.py uses it
}

def updateScripts(scripts):
	## Thanks goes to http://code.activestate.com/recipes/81330-single-pass-multiple-replace/
	from UserDict import UserDict
	import re,os
	class Xlator(UserDict):
		"An all-in-one multiple string substitution class; adapted to match only whole words"
		def _make_regex(self): 
			"Build a regular expression object based on the keys of the current dictionary"
			return re.compile(r"(\b%s\b)" % "|".join(list(self.keys())))  ## adapted here
		def __call__(self, mo):
			"This handler will be invoked for each regex match"
			# Count substitutions
			self.count += 1 # Look-up string
			return self[mo.string[mo.start():mo.end()]]
		def xlat(self, text):
			"Translate text, returns the modified text."
			# Reset substitution counter
			self.count = 0 
			# Process text
			return self._make_regex().sub(self, text)
	# use the _deprecated dictionary for translation, but only when matching on words boundary
	xlator=Xlator(_deprecated)
	if len(scripts)==0: print("No scripts given to --update. Nothing to do.")
	for s in scripts:
		if not s.endswith('.py'): raise RuntimeError("Refusing to do --update on file '"+s+"' (not *.py)")
		txt=open(s).read()
		txt2=xlator.xlat(txt)
		if xlator.count==0: print("%s: already up-to-date."%s)
		else:
			os.rename(s,s+'~')
			out=open(s,'w'); out.write(txt2); out.close()
			print("%s: %d subtitution%s made, backup in %s~"%(s,xlator.count,'s' if xlator.count>1 else '',s))


def cxxCtorsDict(proxyNamespace=__builtins__):
	"""Return dictionary of class constructors for yade's c++ types, which should be used to update a namespace.

	Root classes are those that are directly wrapped by boost::python. These are only put to the dict.

	Derived classes (from these root classes) are faked by creating a callable which invokes appropriate root class constructor with the derived class parameter and passes remaining arguments to it.

	Classes that are neither root nor derived are exposed via callable object that constructs a Serializable of given type and passes the parameters.
	"""
	proxyNamespace={}

	import yade.wrapper
	for c in _allSerializables:
		try:
			proxyNamespace[c]=yade.wrapper.__dict__[c]
		except KeyError: pass # not registered properly

	# deprecated names
	for oldName in list(_deprecated.keys()):
		class warnWrap(object):
			def __init__(self,_old,_new):
				# assert(proxyNamespace.has_key(_new))
				self.old,self.new=_old,_new
			def __call__(self,*args,**kw):
				import warnings; warnings.warn("Class `%s' was renamed to (or replaced by) `%s', update your code! (you can run 'yade --update script.py' to do that automatically)"%(self.old,self.new),DeprecationWarning,stacklevel=2);
				return yade.wrapper.__dict__[self.new](*args,**kw)
		proxyNamespace[oldName]=warnWrap(oldName,_deprecated[oldName])
	return proxyNamespace


## BOOST_LOGGER ## def setExitHandlers():
## BOOST_LOGGER ## 	"""Set exit handler to avoid gdb run if log4cxx crashes at exit."""
## BOOST_LOGGER ## 	# avoid backtrace at regular exit, even if we crash
## BOOST_LOGGER ## 	if 'log4cxx' in config.features:
## BOOST_LOGGER ## 		__builtins__['quit']=wrapper.Omega().exitNoBacktrace
## BOOST_LOGGER ## 		sys.exit=wrapper.Omega().exitNoBacktrace
## BOOST_LOGGER ## 	# this seems to be not needed anymore:
## BOOST_LOGGER ## 	#sys.excepthook=sys.__excepthook__ # apport on ubuntu overrides this, we don't need it


# consistency check
# if there are no serializables, then plugins were not loaded yet, probably
if(len(_allSerializables)==0):
	raise ImportError("No classes deriving from Serializable found; you must call yade.boot.initialize to load plugins before importing yade.system.")

